### PR TITLE
Prettier experiment (discussion only)

### PR DESCRIPTION
**Pull request by octogonz**
Here's an example of how this repo would look after being reformatted using [Prettier](https://prettier.io/).

### CODE REVIEW

**Comment by halfnibble** [comment]
I like it! It's reminiscent of Black for Python. A lot of the styling choices match my preferences, and it cleans up what looks like some obvious formatting "mistakes" too. How difficult is it to implement the tool? Would contributors be required to run it before a PR?

**Comment by octogonz** [comment]
It normally would run automatically via a Git commit hook.  There's some discussion of that in [the Rush docs](https://rushjs.io/pages/maintainer/git_hooks/) but maybe we can improve it.  

I'm also investigating replacing TSLint with the new TypeScript processor for ESLint, in which case Prettier would potentially replace a subset of lint rules dealing with code formatting.

**Comment by sachinjoseph** [comment]
I like this, too! Would this help in IDEs as well? Should we add an editorconfig as well?

**Comment by halfnibble** [comment]
It would be great if this could be added to VSCode. One of the things I loved about Black was that I would never think about formatting. I'd just type until my OCD kicked in, and then I'd hit CTRL+S, and Black would auto-format everything.

**Comment by octogonz** [comment]
See https://github.com/prettier/prettier-vscode :-)

File: apps/api-documenter/src/documenters/MarkdownDocumenter.ts [hunk 1]
```diff
@@ -140,20 +136,17 @@ export class MarkdownDocumenter {
...
-                new DocParagraph({ configuration: this._tsdocConfiguration }, [
-                  new DocPlainText({
-                    configuration: this._tsdocConfiguration,
-                    text: 'Warning: This API is now obsolete. '
-                  })
-                ]),
-                ...tsdocComment.deprecatedBlock.content.nodes
-              ]
-            )
+            new DocNoteBox({ configuration: this._tsdocConfiguration }, [
```
Comment (iclanton) [inline apps/api-documenter/src/documenters/MarkdownDocumenter.ts]: I don't like this change.

File: apps/api-documenter/src/documenters/YamlDocumenter.ts [hunk 2]
```diff
@@ -37,20 +37,14 @@ import {
...
-  IYamlApiFile,
-  IYamlItem,
-  IYamlSyntax,
-  IYamlParameter
-} from '../yaml/IYamlApiFile';
-import {
-  IYamlTocFile,
-  IYamlTocItem
-} from '../yaml/IYamlTocFile';
+import { IYamlApiFile, IYamlItem, IYamlSyntax, IYamlParameter } from '../yaml/IYamlApiFile';
```
Comment (iclanton) [inline apps/api-documenter/src/documenters/YamlDocumenter.ts]: I wish there was an option to always split these.

File: apps/api-documenter/src/documenters/YamlDocumenter.ts [hunk 3]
```diff
@@ -96,12 +90,13 @@ export class YamlDocumenter {
...
       name: 'SharePoint Framework reference',
       href: '~/overview/sharepoint.md',
-      items: [ ]
+      items: []
     };
   }
 
   /** @virtual */
-  protected onCustomizeYamlItem(yamlItem: IYamlItem): void { // virtual
+  protected onCustomizeYamlItem(yamlItem: IYamlItem): void {
```
Comment (iclanton) [inline apps/api-documenter/src/documenters/YamlDocumenter.ts]: Should the `// virtual` comment even be here?

**Comment by iclanton** [comment]
I like the consistency between files, but there are a few things about Prettier's formatting that I really can't stand. The way it doesn't consistently break things by newlines particularly bothers me.

**Comment by octogonz** [comment]
> I like the consistency between files, but there are a few things about Prettier's formatting that I really can't stand. The way it doesn't consistently break things by newlines particularly bothers me.

@iclanton Before we "review" this PR too much, here's a bit of context regarding Prettier (based on my own limited reading and speculation):

The concept of ESLint/TSLint "**fixers**" is that they go into the middle of your file and apply transforms to edit it.  After a bunch of fixers have had their turn, the result is hopefully better.

The concept of Prettier is fundamentally different because it's a **printer** algorithm.  It ignores the human formatting and redoes everything entirely (with a few special exceptions).  This strategy has some technical advantages over **fixers**:
- The **printer** algorithm can consider the document holistically, whereas **fixers** have to be somewhat stupid to avoid unforeseen interactions between fixers
- It's faster, because the printer is a single integrated pass (versus separate passes for each fixer)
- The printer output is in a (more or less) unique "normal form", whereas fixers could converge to different outcomes based on the initial formatting.  As far as heuristics, I imagine it's way easier to tune "the normal form" versus trying to tune all possible inputs the fixers may process.

A consequence is that Prettier really opposes customization.  Its [list of options](https://prettier.io/docs/en/options.html) is paltry.  The docs claim that the "opinionated" stance is a human concern [to avoid wasteful debates](https://prettier.io/docs/en/option-philosophy.html), but I suspect a larger concern is that the **printer** heuristics are very difficult to get right. Every extra option would introduce a bunch of new edge cases to deal with.

Another consequence is that the printer is a difficult algorithm to implement.  Discarding the human formatting means Prettier has full responsibility for the outcome.  The heuristics must simulate what a human would want with reasonable accuracy.  If it gets something wrong, humans have no easy way to fix it.  The Prettier project lives or dies by its heuristics working.  As such, I'd guess that the maintainers should be eager to accept community PRs for **objective** (non-opinion) problems.

With this PR I intended to a discussion of whether to adopt Prettier in Rush Stack; WBT is just a guinea pig.  Our target scenario of "scalable monorepos" involves lots of novice contributors.  Normalizing the look+feel of the source code is a very attractive prospect to me.  The WBT repo today is in unusually good shape because its developers are unusually experienced and diligent. In a typical ecosystem, I wouldn't be surprised if everyday human unprettiness turned out to be worse than the printer algorithm's worst bugs haha.

So, as potential adopters of Prettier, maybe our decision could be framed around questions such as:
- Are Prettier's **opinionated** choices "good enough" for us to accept as our standard? 
- Where Prettier does an **objectively** bad job (e.g. https://github.com/prettier/prettier/issues/5554#issuecomment-509866817), is the issue tolerable?  (In this PR I mostly worked around that one by increasing the `printWidth` option.)
- Where the issue isn't tolerable, do we believe we can solve it by contributing fixes to Prettier?  (Is it easy enough to work on their code?  Would the maintainers accept our PRs?  Is that less cost better than the alternative?)
- If not, what is a viable alternative to choosing Prettier?  I'm not sure TSLint/ESLint fixers are sufficient.

File: apps/api-documenter/src/documenters/YamlDocumenter.ts (hunk 3 above)
Comment (octogonz) [reply apps/api-documenter/src/documenters/YamlDocumenter.ts]: Nope.  `// virtual` was my old convention before TSDoc had `@virtual`. Looks like I forgot to delete it in [`eaf8b5`](https://github.com/microsoft/web-build-tools/commit/eaf8b55659c1112307e3773bb9a1a39d4500ba25#diff-63cb599119dc3cf0dfa99bfa3ecc8f15)

**Comment by apostolisms** [comment]
I agree with @iclanton there are some things that bother me too but possibly tolerable when we can gain overall consistency (?) I guess it is how @octogonz puts it "Are Prettier's opinionated choices "good enough" for us to accept as our standard?"

**Comment by octogonz** [comment]
Here's another option: https://unibeautify.com/

**Comment by octogonz** [comment]
> Here's another option: [unibeautify.com](https://unibeautify.com/)

Heh, it's just a front-end for Prettier.

**Comment by octogonz** [comment]
As an experiment, I spent some time debugging Prettier [issue #5554](https://github.com/prettier/prettier/issues/5554#issuecomment-509866817).  Their code is in untyped JavaScript (boo), but they have a good Jest setup that makes it fairly easy to debug.  Their AST is methodless JSO that can be dumped using `JSON.stringify()`, so you can inspect it. 

It didn't take me very long to track the issue down to this function:

[printer-estree.js](https://github.com/prettier/prettier/blob/033002cdc3cc7e9be1ba8bbea000112373958ea7/src/language-js/printer-estree.js#L4652)
```js
function printTypeParameters(path, options, print, paramsKey) {
  const n = path.getValue();

  if (!n[paramsKey]) {
    return "";
  }

  // for TypeParameterDeclaration typeParameters is a single node
  if (!Array.isArray(n[paramsKey])) {
    return path.call(print, paramsKey);
  }

  const grandparent = path.getNode(2);

  const isParameterInTestCall = grandparent != null && isTestCall(grandparent);

  const shouldInline =
    isParameterInTestCall ||
    n[paramsKey].length === 0 ||
    (n[paramsKey].length === 1 &&
      (shouldHugType(n[paramsKey][0]) ||
        (n[paramsKey][0].type === "GenericTypeAnnotation" &&
          shouldHugType(n[paramsKey][0].id)) ||
        (n[paramsKey][0].type === "TSTypeReference" &&
          shouldHugType(n[paramsKey][0].typeName)) ||
        n[paramsKey][0].type === "NullableTypeAnnotation"));

  if (shouldInline) {
    return concat(["<", join(", ", path.map(print, paramsKey)), ">"]);
  }

  return group(
    concat([
      "<",
      indent(
        concat([
          softline,
          join(concat([",", line]), path.map(print, paramsKey))
        ])
      ),
      ifBreak(
        options.parser !== "typescript" && shouldPrintComma(options, "all")
          ? ","
          : ""
      ),
      softline,
      ">"
    ])
  );
}
```

The `shouldInline` switch decides whether to print this:
```ts
Map<
    string,
    MapItem
  >
```
or this:
```ts
Map<string, MapItem>
```

So I'd imagine the fix would involve tuning the heuristic some more.

Overall I feel like Prettier is a reasonable bet in terms of fixability.

I searched but was not able to find any professional quality TypeScript formatters to contrast with.  If someone has suggestions, that would be appreciated!

**Comment by j-f1** [comment]
Sorry to butt in here, but if you’re planning to contribute and you want to work on the AST, check out [AST explorer](https://astexplorer.net). You’ll want to select `@typescript-eslint/parser` from the `</>` menu to parse TypeScript properly,  and I usually enable “Hide location data” and “Hide type keys” to clean up the AST view so more fits on one screen.

**Comment by wbern** [comment]
Prettier is a great way to unify work. I didn't like the selective ordering of newlines at first, since prettier only prioritizes line width, but after a while it started to make more sense for me, because the code was still more than readable.

**Comment by octogonz** [comment]
BTW [this guide](https://blog.matterhorn.dev/posts/learn-typescript-linting-part-1/) seems to advocate an alternative tool [StandardJS](https://standardjs.com/).  It is a strongly opinionated code formatter based on a bunch of ESLint fixers (including some custom rules).

They [claim](https://standardjs.com/index.html#can-i-use-a-javascript-language-variant-like-flow-or-typescript) to support TypeScript, however the results were pretty unsatisfactory:
- It deleted all the semicolons in our code (and when I tried [semistandard](https://github.com/standard/semistandard) fork, it ignored the TypeScript plugin)
- Otherwise, it didn't fix much of anything beyond indenting the code
- It reports hundreds of `___ is defined but never used` errors because it's clearly not analyzing type annotations

Seems like Prettier doesn't have a lot of competition.

**Comment by halfnibble** [comment]
I'm actually a fan of not using semicolons. :)
I don't know what the downsides are, but I've worked in a JS project without a single semicolon in the src (also avoided traditional for loops), and I liked it.

**Comment by j-f1** [comment]
The only downside to not using semicolons is that if you start a line with one of `` [(`+- ``, it could be interpreted as a continuation of the previous line. Fortunately, if you’re using Prettier, it‘ll point that out for you very clearly, so if that’s what you meant to do, you can leave it, or you can insert a semicolon. Additionally, Prettier will automatically add semicolons before a line starting with any of those things if it interprets it as a statement, to prevent future errors.

**Comment by octogonz** [comment]
I don't see that as a downside really. A linter can catch those mistakes, too.

Far as I can see, the only downside of not using semicolons is that nobody does that. :-P The TypeScript compiler project itself uses semicolons.

In order to break consistency with the majority, we would need to show a compelling upside, not a lack of downsides.

**Comment by iclanton** [comment]
I'm coming around to Prettier, as long as we have a Git hook that runs it.

**Comment by wbern** [comment]
@iclanton we use prettier without a hook today, but we'd love to see how you'd do it. There are some questions on whether the hooks should be managed in each package or if it's a root script. We don't want to impose limitations on our packages so for us it would be nice to have it in each package script.

For setting up hooks you have husky, and also yorkie, somewhat custom made for monorepos.

**Comment by KevinTCoughlin** [comment]
FWIW we've had success integrating Prettier into a few Rush repos using the documentation at https://rushjs.io/pages/maintainer/git_hooks/

**Comment by octogonz** [comment]
> For setting up hooks you have husky, and also yorkie, somewhat custom made for monorepos.

FYI Rush has [built-in support](https://rushjs.io/pages/maintainer/git_hooks/) for configuring Git hooks.

We tried Husky, but we had issues because it assumes that it can find a **package.json** file by probing upwards from its installation folder, which is only true for the NPM-style **node_modules** layout.  Also Husky creates a Node process for every Git event (even unused hooks) which is not very efficient.  I wasn't aware of **yorkie** -- maybe it solves those problems.

Rush's approach does require you to write Bash scripts instead of `npm run` scripts, but Bash isn't too difficult to learn.

**Comment by halfnibble** [comment]
I’m starting to see some traction in the “no semicolons” camp. Boris Cherny’s new “Programming TypeScript” book uses no optional semicolons throughout. (It’s a great read too, btw).

**Comment by octogonz** [comment]
> we use prettier without a hook today, but we'd love to see how you'd do it. There are some questions on whether the hooks should be managed in each package or if it's a root script. We don't want to impose limitations on our packages so for us it would be nice to have it in each package script, I think..

I've been thinking a whole lot about this problem.  Prettier and ESLint have a very similar design, so the considerations apply equally to them.

In a Rush repo, if you have hundreds of projects, each project should have its own linting configuration, which can extend its toolchain default.  Some projects won't get linted at all.  Or they may have files (e.g. test assets) that look like source code but should not be linted.  Or during a migration, some projects may use a different ESLint release.  Thus normally we would expect to treat **linting** as a typical project build step, and **fixup** as a project task that can be invoked manually.

Unfortunately Prettier and ESLint both seem to assume their CLI gets invoked globally on a relatively small repo.  Specifically:
- The `.prettierignore` and `.eslintignore` files are the only way to specify which files to process, but there's no way to share them between projects.  
- The processing of these files is inefficient; ignoring an entire subfolder does not seem to prevent the **file-enumerator** from recursively testing every single path against the globs.  It's probably acceptable for a single invocation in a small repo, but imagine running this 1000 times in a large monorepo.
- Both Prettier and ESLint config files are missing important settings that can only be specified as CLI options; this means the VS Code extension won't work right without special VS Code settings for each project (e.g. `--ext .ts --ext .tsx` in ESLint's case)

Supposing we get project-specific configs working, a separate question arises for a Git commit hook like [pretty-quick](https://github.com/azz/pretty-quick).  During commit, it wants to perform fixup for only the lines of code affected by your Git diff.  How would this work with project-specific configurations?  The hook would need to compute the subset of affected Rush projects, and then for each project, it would intersect the diff with the project's globs and invoke project-specific fixup.

All these problems are solvable.  But it seems like maybe Rush and our toolchain need to be more involved with how the linters get invoked.

**Comment by wbern** [comment]
>All these problems are solvable. But it seems like maybe Rush and our toolchain need to be more involved with how the linters get invoked.

That sounds like a key ingredient for rush indeed. We have prettier today, and I think it may respect jsconfig.json. We had extremely slow linting with Vetur, a vscode extension that runs prettier and eslint on vue files, before adding a good jsconfig file.

I also think prettier has a pretty bad cli. It would be nice if it could be improved.

**Comment by j-f1** [comment]
Here’s an option: https://github.com/azz/pretty-quick

**Comment by octogonz** [comment]
> Discarding the human formatting means Prettier has full responsibility for the outcome. The heuristics must simulate what a human would want with reasonable accuracy. If it gets something wrong, humans can't just edit their code to undo it.

Turns out that there is a `// prettier-ignore` directive that applies to the next "node" (i.e. grammatical expression) in the file.  Seems like that's an easy workaround for cases where the autoformatting isn't good.

**Comment by octogonz** [comment]
I'm closing this discussion-only PR.  

Prettier is coming soon. 😊

**Comment by karlhorky** [comment]
@octogonz Is there anywhere we can follow along with this migration? Or will all pull requests tag this PR so that they are linked in the discussion here?

**Comment by octogonz** [comment]
@karlhorky Good point.  I opened https://github.com/microsoft/rushstack/issues/1573 to track this.

**Comment by karlhorky** [comment]
Thanks!

**Pull request closed without merging**